src/Tools/dotnet-monitor/StartupHook/StartupHookService.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.StartupHook
+{
+    internal sealed class StartupHookService :
+        IDiagnosticLifetimeService
+    {
+        private readonly IEndpointInfo _endpointInfo;
+        private readonly IInProcessFeatures _inProcessFeatures;
+        private readonly TaskCompletionSource<bool> _resultSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+        private readonly StartupHookValidator _startupHookValidator;
+
+        public StartupHookService(
+            IEndpointInfo endpointInfo,
+            IInProcessFeatures inProcessFeatures,
+            StartupHookValidator startupHookValidator)
+        {
+            _endpointInfo = endpointInfo ?? throw new ArgumentNullException(nameof(endpointInfo));
+            _inProcessFeatures = inProcessFeatures ?? throw new ArgumentNullException(nameof(inProcessFeatures));
+            _startupHookValidator = startupHookValidator ?? throw new ArgumentNullException(nameof(startupHookValidator));
+        }
+
+        public async ValueTask StartAsync(CancellationToken cancellationToken)
+        {
+            if (_inProcessFeatures.IsStartupHookRequired)
+            {
+                if (await _startupHookValidator.CheckEnvironmentAsync(_endpointInfo, cancellationToken))
+                {
+                    _resultSource.SetResult(true);
+                    return;
+                }
+
+                if (await _startupHookValidator.ApplyStartupHook(_endpointInfo, cancellationToken))
```

Comment:
I know this is not related to your PR, but I believe we do not properly merge startup hooks, but replace existing ones.

Reply:
We don't directly modify startup hooks. We use the diagnostic IPC channel to apply them. It does appendation within the runtime itself.